In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

In [2]:
data = pd.read_csv('data/GDP_Country.csv')
data.sample(5)

Country  Population  Area_sqm  Pop_Density_per sqm  \
137         Moldova      4466706     33843                132.0   
180         Senegal     11987121    196190                 61.1   
84        Guatemala     12293545    108890                112.9   
213  United Kingdom     60609153    244820                247.6   
60            Egypt     78887007   1001450                 78.8   

     Coastline (coast/area ratio)  Net migration  \
137                          0.00          -0.26   
180                          0.27           0.20   
84                           0.37          -1.67   
213                          5.08           2.19   
60                           0.24          -0.22   

     Infant mortality (per 1000 births)  GDP ($ per capita)  Literacy (%)  \
137                               40.42              1800.0          99.1   
180                               55.51              1600.0          40.2   
84                                35.93              4100.0          70.6   
213                                5.16             27700.0          99.0   
60                                32.59              4000.0          57.7   

     Phones (per 1000)  Arable (%)  Crops (%)  Other (%)  Climate  Birthrate  \
137              208.1       55.30      10.79      33.91      NaN      15.70   
180               22.2       12.78       0.21      87.01      2.0      32.78   
84                92.1       12.54       5.03      82.43      2.0      29.88   
213              543.5       23.46       0.21      76.33      3.0      10.71   
60               131.8        2.87       0.48      96.65      1.0      22.94   

     Deathrate  Agriculture  Industry  Service  
137      12.64        0.213     0.233    0.555  
180       9.42        0.172     0.209    0.619  
84        5.20        0.227     0.188    0.585  
213      10.13        0.005     0.237    0.758  
60        5.23        0.149     0.357    0.493

### Preprocessing

In [3]:
data.isnull().sum()

Country                                0
Population                             0
Area_sqm                               0
Pop_Density_per sqm                    0
Coastline (coast/area ratio)           0
Net migration                          3
Infant mortality (per 1000 births)     3
GDP ($ per capita)                     1
Literacy (%)                          18
Phones (per 1000)                      4
Arable (%)                             2
Crops (%)                              2
Other (%)                              2
Climate                               22
Birthrate                              3
Deathrate                              4
Agriculture                           15
Industry                              16
Service                               15
dtype: int64

In [4]:
data.isnull().sum()/len(data)

Country                               0.000000
Population                            0.000000
Area_sqm                              0.000000
Pop_Density_per sqm                   0.000000
Coastline (coast/area ratio)          0.000000
Net migration                         0.013216
Infant mortality (per 1000 births)    0.013216
GDP ($ per capita)                    0.004405
Literacy (%)                          0.079295
Phones (per 1000)                     0.017621
Arable (%)                            0.008811
Crops (%)                             0.008811
Other (%)                             0.008811
Climate                               0.096916
Birthrate                             0.013216
Deathrate                             0.017621
Agriculture                           0.066079
Industry                              0.070485
Service                               0.066079
dtype: float64

### Imputation

In [5]:
# KNN Imputer doesn't work on Categorical data.
# Country is categorical element, since its only one its set to index
data1 = data.set_index('Country')
data1.sample(2)

Population  Area_sqm  Pop_Density_per sqm  \
Country                                                    
Saudi Arabia     27019731   1960582                 13.8   
Luxembourg         474413      2586                183.5   

               Coastline (coast/area ratio)  Net migration  \
Country                                                      
Saudi Arabia                           0.13          -2.71   
Luxembourg                             0.00           8.97   

               Infant mortality (per 1000 births)  GDP ($ per capita)  \
Country                                                                 
Saudi Arabia                                13.24             11800.0   
Luxembourg                                   4.81             55100.0   

               Literacy (%)  Phones (per 1000)  Arable (%)  Crops (%)  \
Country                                                                 
Saudi Arabia           78.8              140.6        1.67       0.09   
Luxembourg            100.0              515.4       23.28       0.40   

               Other (%)  Climate  Birthrate  Deathrate  Agriculture  \
Country                                                                
Saudi Arabia       98.24      1.0      29.34       2.58        0.033   
Luxembourg         76.32      NaN      11.94       8.41        0.010   

               Industry  Service  
Country                           
Saudi Arabia      0.613    0.354  
Luxembourg        0.130    0.860

In [6]:
from sklearn.impute import KNNImputer

In [7]:
# Alternatively we can use KNNImputer from missingpy
# from missingpy import KNNImputer

In [8]:
impu = KNNImputer()
data2 = impu.fit_transform(data1)
data2 = pd.DataFrame(data2, index=data1.index, columns=data1.columns)

In [9]:
data2.sample(2)

Population  Area_sqm  Pop_Density_per sqm  \
Country                                                            
Guinea                  9690222.0  245857.0                 39.4   
United Arab Emirates    2602713.0   82880.0                 31.4   

                       Coastline (coast/area ratio)  Net migration  \
Country                                                              
Guinea                                         0.13          -3.06   
United Arab Emirates                           1.59           1.03   

                       Infant mortality (per 1000 births)  GDP ($ per capita)  \
Country                                                                         
Guinea                                              90.37              2100.0   
United Arab Emirates                                14.51             23200.0   

                       Literacy (%)  Phones (per 1000)  Arable (%)  Crops (%)  \
Country                                                                         
Guinea                         35.9                2.7        3.63       2.58   
United Arab Emirates           77.9              475.3        0.60       2.25   

                       Other (%)  Climate  Birthrate  Deathrate  Agriculture  \
Country                                                                        
Guinea                     93.79      2.0      41.76      15.48        0.237   
United Arab Emirates       97.15      1.0      18.96       4.40        0.040   

                       Industry  Service  
Country                                   
Guinea                    0.362    0.401  
United Arab Emirates      0.585    0.375

In [10]:
data2.isnull().sum()

Population                            0
Area_sqm                              0
Pop_Density_per sqm                   0
Coastline (coast/area ratio)          0
Net migration                         0
Infant mortality (per 1000 births)    0
GDP ($ per capita)                    0
Literacy (%)                          0
Phones (per 1000)                     0
Arable (%)                            0
Crops (%)                             0
Other (%)                             0
Climate                               0
Birthrate                             0
Deathrate                             0
Agriculture                           0
Industry                              0
Service                               0
dtype: int64

### Base regression Model
Later we can do the outlier treatment to check if the model has improved

In [11]:
out = data2['GDP ($ per capita)']
inp = data2.drop('GDP ($ per capita)', 1)

In [12]:
import statsmodels.api as sm

In [13]:
inpc = sm.add_constant(inp)
ols = sm.OLS(out, inpc)
ols_mod = ols.fit()
ols_mod.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     GDP ($ per capita)   R-squared:                       0.754
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     37.61
Date:                Fri, 03 Sep 2021   Prob (F-statistic):           4.32e-54
Time:                        07:50:03   Log-Likelihood:                -2254.0
No. Observations:                 227   AIC:                             4544.
Df Residuals:                     209   BIC:                             4606.
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
const                               6.946e+06   4.41e+06      1.576      0.117   -1.74e+06    1.56e+07
Population                         -2.173e-06   3.51e-06     -0.619      0.537   -9.09e-06    4.75e-06
Area_sqm                            7.452e-05      0.000      0.323      0.747      -0.000       0.001
Pop_Density_per sqm                   -0.2639      0.230     -1.149      0.252      -0.717       0.189
Coastline (coast/area ratio)          -1.1338      5.545     -0.204      0.838     -12.065       9.797
Net migration                        416.8470     78.894      5.284      0.000     261.317     572.377
Infant mortality (per 1000 births)   -37.9143     28.409     -1.335      0.183     -93.918      18.090
Literacy (%)                          -3.6942     30.549     -0.121      0.904     -63.918      56.530
Phones (per 1000)                     30.3023      2.688     11.274      0.000      25.003      35.601
Arable (%)                         -6.936e+04   4.41e+04     -1.574      0.117   -1.56e+05    1.75e+04
Crops (%)                          -6.936e+04   4.41e+04     -1.574      0.117   -1.56e+05    1.75e+04
Other (%)                          -6.934e+04   4.41e+04     -1.573      0.117   -1.56e+05    1.75e+04
Climate                               -2.9871    652.319     -0.005      0.996   -1288.956    1282.982
Birthrate                            -58.5084     76.967     -0.760      0.448    -210.240      93.223
Deathrate                            172.1638    109.059      1.579      0.116     -42.834     387.161
Agriculture                        -1.017e+04   1.55e+04     -0.658      0.511   -4.07e+04    2.03e+04
Industry                           -3066.6365   1.51e+04     -0.203      0.839   -3.28e+04    2.66e+04
Service                            -8334.1693   1.51e+04     -0.550      0.583   -3.82e+04    2.15e+04
==============================================================================
Omnibus:                       78.912   Durbin-Watson:                   1.828
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              611.087
Skew:                           1.115   Prob(JB):                    2.01e-133
Kurtosis:                      10.722   Cond. No.                     1.55e+12
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.55e+12. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Intepretation

Date:	Thu, 02 Sep 2021	Prob (F-statistic):	4.32e-54

Since F-Stat < 0.05 the overall model is significant. H0 is rejected in favour of Ha

Individual features some of them are greater than 0.05, we will start looking at feature with highest p-value and drop it, again re-build the model and check again p-values 

### Outliear Treatment

In [14]:
# data2[data2['Area_sqm']>data2['Area_sqm'].quantile(.75)]['Area_sqm']

In [15]:
# data2.fillna(0)
# data2.isnull().sum()

In [16]:
# for i in data2.columns:
#     q1 = data2[i].quantile(.25)
#     q3 = data2[i].quantile(.75)
#     iqr = q3 - q1
#     ub = q3 + 1.5 * iqr
#     lb = q1 - 1.5 * iqr
#     data2[i] = data2[i].replace(to_replace=data2[data2[i]>ub][i], value=data2[i].quantile(0.99))
#     data2[i] = data2[i].replace(to_replace=data2[data2[i]<lb][i], value=data2[i].quantile(0.01))

In [17]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 227 entries, Afghanistan  to Zimbabwe 
Data columns (total 18 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Population                          227 non-null    float64
 1   Area_sqm                            227 non-null    float64
 2   Pop_Density_per sqm                 227 non-null    float64
 3   Coastline (coast/area ratio)        227 non-null    float64
 4   Net migration                       227 non-null    float64
 5   Infant mortality (per 1000 births)  227 non-null    float64
 6   GDP ($ per capita)                  227 non-null    float64
 7   Literacy (%)                        227 non-null    float64
 8   Phones (per 1000)                   227 non-null    float64
 9   Arable (%)                          227 non-null    float64
 10  Crops (%)                           227 non-null    float64
 11  Other (%)                        

In [18]:
for i in data2.columns:
    q1 = data2[i].quantile(.25)
    q3 = data2[i].quantile(.75)
    iqr = q3 - q1
    ub = q3 + 1.5 * iqr
    lb = q1 - 1.5 * iqr
    data2[i] = np.where(data2[i]>ub, 
         data2[i].quantile(.99), 
        np.where(data2[i]<lb, 
                 data2[i].quantile(.01),
                data2[i]))


In [19]:
data2.sample(5)

Population   Area_sqm  Pop_Density_per sqm  \
Country                                                            
Jersey                   91084.0      116.0              6449.31   
Sao Tome & Principe     193413.0     1001.0               193.20   
Libya                  5900754.0  9622460.4                 3.40   
Greece                10688058.0   131940.0                81.00   
Palau                    20579.0      458.0                44.90   

                      Coastline (coast/area ratio)  Net migration  \
Country                                                             
Jersey                                    285.7248           2.76   
Sao Tome & Principe                        20.8800          -2.72   
Libya                                       0.1000           0.00   
Greece                                     10.3700           2.35   
Palau                                     285.7248           2.85   

                      Infant mortality (per 1000 births)  GDP ($ per capita)  \
Country                                                                        
Jersey                                              5.24             24800.0   
Sao Tome & Principe                                43.11              1200.0   
Libya                                              24.60              6400.0   
Greece                                              5.53             20000.0   
Palau                                              14.84              9000.0   

                      Literacy (%)  Phones (per 1000)  Arable (%)  Crops (%)  \
Country                                                                        
Jersey                        90.0              811.3        0.00      0.000   
Sao Tome & Principe           79.3               36.2        6.25     45.021   
Libya                         82.6              127.1        1.03      0.190   
Greece                        97.5              589.7       21.10      8.780   
Palau                         92.0              325.6        8.70      4.350   

                      Other (%)  Climate  Birthrate  Deathrate  Agriculture  \
Country                                                                       
Jersey                   100.00      3.0       9.30       9.28        0.050   
Sao Tome & Principe       44.79      2.0      40.25       6.47        0.167   
Libya                     98.78      2.0      26.49       3.48        0.076   
Greece                    70.12      3.0       9.68      10.24        0.054   
Palau                     86.95      2.0      18.03       6.80        0.062   

                      Industry  Service  
Country                                  
Jersey                   0.020    0.930  
Sao Tome & Principe      0.148    0.684  
Libya                    0.499    0.425  
Greece                   0.213    0.733  
Palau                    0.120    0.818

In [20]:
out = data2['GDP ($ per capita)']
inp = data2.drop('GDP ($ per capita)', axis=1)

In [21]:
inpc = sm.add_constant(inp)
ols = sm.OLS(out, inpc)
ols_mod = ols.fit()
ols_mod.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     GDP ($ per capita)   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.785
Method:                 Least Squares   F-statistic:                     49.48
Date:                Fri, 03 Sep 2021   Prob (F-statistic):           1.40e-63
Time:                        07:50:03   Log-Likelihood:                -2222.8
No. Observations:                 227   AIC:                             4482.
Df Residuals:                     209   BIC:                             4543.
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
const                               8642.6765   1.21e+04      0.715      0.475   -1.52e+04    3.25e+04
Population                          1.698e-06   3.89e-06      0.437      0.663   -5.97e-06    9.36e-06
Area_sqm                           -1.905e-05      0.000     -0.166      0.868      -0.000       0.000
Pop_Density_per sqm                    0.0281      0.188      0.150      0.881      -0.343       0.399
Coastline (coast/area ratio)         -15.1276      3.853     -3.926      0.000     -22.723      -7.532
Net migration                        295.1782     45.856      6.437      0.000     204.779     385.577
Infant mortality (per 1000 births)   -37.6590     26.997     -1.395      0.165     -90.881      15.563
Literacy (%)                          -0.4939     25.102     -0.020      0.984     -49.979      48.991
Phones (per 1000)                     32.2409      2.414     13.358      0.000      27.483      36.999
Arable (%)                           -75.8229     72.753     -1.042      0.299    -219.247      67.601
Crops (%)                            -21.7611     39.638     -0.549      0.584     -99.903      56.381
Other (%)                            -34.6986     70.617     -0.491      0.624    -173.912     104.515
Climate                             -187.4946    561.108     -0.334      0.739   -1293.652     918.663
Birthrate                            -30.8857     66.935     -0.461      0.645    -162.839     101.068
Deathrate                            125.3649     80.701      1.553      0.122     -33.726     284.456
Agriculture                        -4532.7267   9322.350     -0.486      0.627   -2.29e+04    1.38e+04
Industry                            2115.0018   8947.628      0.236      0.813   -1.55e+04    1.98e+04
Service                            -1155.9880   9240.786     -0.125      0.901   -1.94e+04    1.71e+04
==============================================================================
Omnibus:                        9.555   Durbin-Watson:                   1.855
Prob(Omnibus):                  0.008   Jarque-Bera (JB):               15.716
Skew:                           0.209   Prob(JB):                     0.000387
Kurtosis:                       4.220   Cond. No.                     6.21e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.21e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Scaling
Bring data into a defined range

In [22]:
from sklearn.preprocessing import StandardScaler

In [23]:
sc = StandardScaler()
inp_sc = sc.fit_transform(inp)
inp_sc = pd.DataFrame(inp_sc, columns=inp.columns, index=inp.index)
inp_sc.head(5)

Population  Area_sqm  Pop_Density_per sqm  \
Country                                                      
Afghanistan       -0.115094 -0.220266            -0.347864   
Albania           -0.414949 -0.418770            -0.306272   
Algeria           -0.094652  2.659029            -0.366433   
American Samoa    -0.453407 -0.427929            -0.216248   
Andorra           -0.453260 -0.427843            -0.291340   

                 Coastline (coast/area ratio)  Net migration  \
Country                                                        
Afghanistan                         -0.470180       2.143829   
Albania                             -0.458127      -1.993115   
Algeria                             -0.469797      -0.107041   
American Samoa                       2.263057      -1.993115   
Andorra                             -0.470180       2.143829   

                 Infant mortality (per 1000 births)  Literacy (%)  \
Country                                                             
Afghanistan                                3.079871     -2.537552   
Albania                                   -0.393970      0.181254   
Algeria                                   -0.116716     -0.650571   
American Samoa                            -0.752237      0.710596   
Andorra                                   -0.904902      0.861837   

                 Phones (per 1000)  Arable (%)  Crops (%)  Other (%)  \
Country                                                                
Afghanistan              -1.041107   -0.132135  -0.502354   0.375437   
Albania                  -0.736564    0.539398  -0.218337  -0.449624   
Algeria                  -0.705662   -0.799921  -0.500325   0.932166   
American Samoa            0.106750   -0.291774   2.527233  -0.417649   
Andorra                   1.171305   -0.874869  -0.517231   1.010534   

                  Climate  Birthrate  Deathrate  Agriculture  Industry  \
Country                                                                  
Afghanistan     -1.694570   2.210162   3.055247     1.612835 -0.263552   
Albania          1.260999  -0.631504  -0.738826     0.583345 -0.645495   
Algeria         -1.694570  -0.448316  -0.842529    -0.327892  2.850162   
American Samoa  -0.216785   0.031762  -1.070336    -0.199902 -0.476559   
Andorra          1.260999  -1.209041  -0.563721    -0.670128 -0.394294   

                  Service  
Country                    
Afghanistan     -1.201003  
Albania          0.014199  
Algeria         -1.701740  
American Samoa   0.568673  
Andorra          0.913082

In [24]:
# There is no change after scaling the data
inp['Population'].skew(), inp_sc["Population"].skew()

(2.2625680522702516, 2.262568052270252)

In [25]:
inpc = sm.add_constant(inp_sc)
ols = sm.OLS(out, inpc)
ols_mod = ols.fit()
ols_mod.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     GDP ($ per capita)   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.785
Method:                 Least Squares   F-statistic:                     49.48
Date:                Fri, 03 Sep 2021   Prob (F-statistic):           1.40e-63
Time:                        07:50:03   Log-Likelihood:                -2222.8
No. Observations:                 227   AIC:                             4482.
Df Residuals:                     209   BIC:                             4543.
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
const                               9620.3348    299.480     32.123      0.000    9029.946    1.02e+04
Population                           155.5794    356.182      0.437      0.663    -546.590     857.749
Area_sqm                             -59.3790    357.475     -0.166      0.868    -764.099     645.341
Pop_Density_per sqm                   51.8321    346.271      0.150      0.881    -630.799     734.463
Coastline (coast/area ratio)       -1581.3918    402.751     -3.926      0.000   -2375.366    -787.417
Net migration                       2115.4650    328.636      6.437      0.000    1467.598    2763.331
Infant mortality (per 1000 births) -1287.6527    923.107     -1.395      0.165   -3107.447     532.142
Literacy (%)                          -9.7969    497.915     -0.020      0.984    -991.376     971.782
Phones (per 1000)                   7198.9293    538.933     13.358      0.000    6136.488    8261.371
Arable (%)                         -1011.6755    970.714     -1.042      0.299   -2925.321     901.970
Crops (%)                           -321.7995    586.161     -0.549      0.584   -1477.345     833.746
Other (%)                           -553.4537   1126.368     -0.491      0.624   -2773.953    1667.046
Climate                             -126.8755    379.696     -0.334      0.739    -875.400     621.649
Birthrate                           -342.2609    741.738     -0.461      0.645   -1804.507    1119.985
Deathrate                            737.4182    474.695      1.553      0.122    -198.385    1673.221
Agriculture                         -651.6274   1340.186     -0.486      0.627   -3293.643    1990.389
Industry                             287.9492   1218.184      0.236      0.813   -2113.554    2689.453
Service                             -189.3032   1513.260     -0.125      0.901   -3172.512    2793.906
==============================================================================
Omnibus:                        9.555   Durbin-Watson:                   1.855
Prob(Omnibus):                  0.008   Jarque-Bera (JB):               15.716
Skew:                           0.209   Prob(JB):                     0.000387
Kurtosis:                       4.220   Cond. No.                         17.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [26]:
print("There is not change before scaling and after scaling")

There is not change before scaling and after scaling


##### Interpretation

Before Standarization:

Kurtosis:	4.220	Cond. No.	6.21e+09

After Standarization:

Kurtosis:	4.220	Cond. No.	17.5

If the Condition Number is small or less than 100 that means that multicollinearity effect is very small. VIF gives the exact but this Coef No is an inference.

Standization reduced the Condition number i.e., reduces the multicollinearity problem

If we look at coef there is a change in coef. The inference will change since the coef has changed



In [27]:
# Instead of StandardScaler just do the mean subtraction
inp1 = inp - inp.mean()
inpc = sm.add_constant(inp1)
ols = sm.OLS(out, inpc)
ols_mod = ols.fit()
ols_mod.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     GDP ($ per capita)   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.785
Method:                 Least Squares   F-statistic:                     49.48
Date:                Fri, 03 Sep 2021   Prob (F-statistic):           1.40e-63
Time:                        07:50:03   Log-Likelihood:                -2222.8
No. Observations:                 227   AIC:                             4482.
Df Residuals:                     209   BIC:                             4543.
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
const                               9620.3348    299.480     32.123      0.000    9029.946    1.02e+04
Population                          1.698e-06   3.89e-06      0.437      0.663   -5.97e-06    9.36e-06
Area_sqm                           -1.905e-05      0.000     -0.166      0.868      -0.000       0.000
Pop_Density_per sqm                    0.0281      0.188      0.150      0.881      -0.343       0.399
Coastline (coast/area ratio)         -15.1276      3.853     -3.926      0.000     -22.723      -7.532
Net migration                        295.1782     45.856      6.437      0.000     204.779     385.577
Infant mortality (per 1000 births)   -37.6590     26.997     -1.395      0.165     -90.881      15.563
Literacy (%)                          -0.4939     25.102     -0.020      0.984     -49.979      48.991
Phones (per 1000)                     32.2409      2.414     13.358      0.000      27.483      36.999
Arable (%)                           -75.8229     72.753     -1.042      0.299    -219.247      67.601
Crops (%)                            -21.7611     39.638     -0.549      0.584     -99.903      56.381
Other (%)                            -34.6986     70.617     -0.491      0.624    -173.912     104.515
Climate                             -187.4946    561.108     -0.334      0.739   -1293.652     918.663
Birthrate                            -30.8857     66.935     -0.461      0.645    -162.839     101.068
Deathrate                            125.3649     80.701      1.553      0.122     -33.726     284.456
Agriculture                        -4532.7267   9322.350     -0.486      0.627   -2.29e+04    1.38e+04
Industry                            2115.0018   8947.628      0.236      0.813   -1.55e+04    1.98e+04
Service                            -1155.9880   9240.786     -0.125      0.901   -1.94e+04    1.71e+04
==============================================================================
Omnibus:                        9.555   Durbin-Watson:                   1.855
Prob(Omnibus):                  0.008   Jarque-Bera (JB):               15.716
Skew:                           0.209   Prob(JB):                     0.000387
Kurtosis:                       4.220   Cond. No.                     4.75e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.75e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

##### Interpretation
Other feature coef remain the same as Pre-scaled data only Constant has change.

Cond. No has changed (down) from the Pre-scaled data


### Linear Regression: Assumptions


Assumptions of Linear Regression:

	1. Multicollinearity:
		a. This is a check for the data
        
	2. Normality:
		a. This is a check for the model
        
	3. Linearity:
		a. This is a check for the model
        
	4. Autocorrelation:
		a. This is a check for the model
        
	5. Homoscedasticity:
		a. This is a check for the model
       



### Multicollinearity

In [28]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [31]:
# Convert dataframe into np array
# VIF is computed using x1 against x2, x3, x4, x5. 
# Next x2 againts x1, x3, x4, x5 iteratively for each of the features against all other features 
vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(inp1.values, i) for i in range(inp1.shape[1])]
vif['feature'] = inp1.columns
vif.sort_values('VIF', ascending=False)

VIF                             feature
16  25.532397                             Service
14  20.026042                         Agriculture
15  16.545912                            Industry
10  14.145739                           Other (%)
8   10.506242                          Arable (%)
5    9.500994  Infant mortality (per 1000 births)
12   6.134302                           Birthrate
9    3.830876                           Crops (%)
7    3.238427                   Phones (per 1000)
6    2.764233                        Literacy (%)
13   2.512426                           Deathrate
3    1.808579        Coastline (coast/area ratio)
11   1.607444                             Climate
1    1.424810                            Area_sqm
0    1.414517                          Population
2    1.336890                 Pop_Density_per sqm
4    1.204191                       Net migration

##### Interpretation
For these features VIF > 5, we will drop one by one feature and then check for 
multicollinearity between other features

VIF	feature

16	25.532397	Service

14	20.026042	Agriculture

15	16.545912	Industry

10	14.145739	Other (%)

8	10.506242	Arable (%)

5	9.500994	Infant mortality (per 1000 births)

12	6.134302	Birthrate

Sometimes Service might be very critial for explainability of the model, so we 
can't drop Service. Lets check the correlation of all the features with respect 
GDP data. Since Industry has less correlation lets try droping the Industry and checking

Agriculture                          -0.580424

Industry                             -0.036918

Service                               0.535850


In [32]:
data2.corr()['GDP ($ per capita)']

Population                            0.017884
Area_sqm                             -0.060373
Pop_Density_per sqm                   0.238214
Coastline (coast/area ratio)          0.177843
Net migration                         0.403517
Infant mortality (per 1000 births)   -0.619008
GDP ($ per capita)                    1.000000
Literacy (%)                          0.513722
Phones (per 1000)                     0.847224
Arable (%)                            0.014793
Crops (%)                            -0.230128
Other (%)                             0.105637
Climate                               0.318543
Birthrate                            -0.661537
Deathrate                            -0.219528
Agriculture                          -0.580424
Industry                             -0.036918
Service                               0.535850
Name: GDP ($ per capita), dtype: float64

In [38]:
inpv1 = inp1.drop('Industry', axis=1)

vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(np.array(inpv1), i) for i in range(inpv1.shape[1])]
vif['feature'] = inpv1.columns
vif.sort_values(by='VIF', ascending=False)

VIF                             feature
10  14.144188                           Other (%)
8   10.504069                          Arable (%)
5    9.500895  Infant mortality (per 1000 births)
12   6.127679                           Birthrate
9    3.830596                           Crops (%)
7    3.238427                   Phones (per 1000)
6    2.742160                        Literacy (%)
15   2.645795                             Service
14   2.570655                         Agriculture
13   2.510800                           Deathrate
3    1.807521        Coastline (coast/area ratio)
11   1.594188                             Climate
0    1.412871                          Population
1    1.404079                            Area_sqm
2    1.332839                 Pop_Density_per sqm
4    1.196891                       Net migration

In [34]:
np.array(inpv1)

array([[-1.05459775e+07, -6.86583005e+05, -6.40668767e+02, ...,
         1.79714952e+01,  2.31862026e-01, -1.96674802e-01],
       [-3.80213195e+07, -1.30533501e+06, -5.64068767e+02, ...,
        -4.34590485e+00,  8.38620264e-02,  2.32519824e-03],
       [-8.67288349e+06,  8.28837739e+06, -6.74868767e+02, ...,
        -4.95590485e+00, -4.71379736e-02, -2.78674802e-01],
       ...,
       [-2.01467865e+07, -8.06113005e+05, -6.48068767e+02, ...,
        -1.26590485e+00, -1.31379736e-02, -1.83674802e-01],
       [-3.01009645e+07, -5.81469005e+05, -6.73368767e+02, ...,
         1.79714952e+01,  7.18620264e-02, -8.76748018e-02],
       [-2.93661695e+07, -9.43503005e+05, -6.57368767e+02, ...,
         1.79714952e+01,  3.08620264e-02,  2.32519824e-03]])

In [35]:
inpv1.values

array([[-1.05459775e+07, -6.86583005e+05, -6.40668767e+02, ...,
         1.79714952e+01,  2.31862026e-01, -1.96674802e-01],
       [-3.80213195e+07, -1.30533501e+06, -5.64068767e+02, ...,
        -4.34590485e+00,  8.38620264e-02,  2.32519824e-03],
       [-8.67288349e+06,  8.28837739e+06, -6.74868767e+02, ...,
        -4.95590485e+00, -4.71379736e-02, -2.78674802e-01],
       ...,
       [-2.01467865e+07, -8.06113005e+05, -6.48068767e+02, ...,
        -1.26590485e+00, -1.31379736e-02, -1.83674802e-01],
       [-3.01009645e+07, -5.81469005e+05, -6.73368767e+02, ...,
         1.79714952e+01,  7.18620264e-02, -8.76748018e-02],
       [-2.93661695e+07, -9.43503005e+05, -6.57368767e+02, ...,
         1.79714952e+01,  3.08620264e-02,  2.32519824e-03]])

##### Interpretation

VIF	feature

10	14.144188	Other (%)

8	10.504069	Arable (%)

5	9.500895	Infant mortality (per 1000 births)

Since Arable has less correlation lets try droping the Arable and checking

In [39]:
inpv2 = inpv1.drop('Arable (%)', axis=1)
vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(np.array(inpv2), i) for i in range(inpv2.shape[1])]
vif['feature'] = inpv2.columns
vif.sort_values('VIF', ascending=False)

VIF                             feature
5   9.482559  Infant mortality (per 1000 births)
11  6.004283                           Birthrate
7   3.235938                   Phones (per 1000)
6   2.705239                        Literacy (%)
14  2.645760                             Service
13  2.570643                         Agriculture
12  2.505773                           Deathrate
9   1.903902                           Other (%)
8   1.885588                           Crops (%)
3   1.731842        Coastline (coast/area ratio)
10  1.587820                             Climate
0   1.412517                          Population
1   1.404061                            Area_sqm
2   1.315430                 Pop_Density_per sqm
4   1.185134                       Net migration

In [40]:
inpv3 = inpv2.drop('Infant mortality (per 1000 births)', axis=1)
vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(exog=np.array(inpv3), exog_idx=i) for i in range(inpv3.shape[1])]
vif['feature'] = inpv3.columns
vif.sort_values('VIF', ascending=False)

VIF                       feature
10  4.430107                     Birthrate
6   3.169817             Phones (per 1000)
13  2.562855                       Service
5   2.562702                  Literacy (%)
12  2.359005                   Agriculture
8   1.901241                     Other (%)
7   1.865200                     Crops (%)
3   1.723922  Coastline (coast/area ratio)
9   1.581554                       Climate
11  1.455762                     Deathrate
1   1.402826                      Area_sqm
0   1.398767                    Population
2   1.313317           Pop_Density_per sqm
4   1.184030                 Net migration

##### Interpretation
Now all the VIF factors are less than 5.
Multicollinearity problem is addressed by droping 
'Industry', 'Arable (%)', 'Infant mortality (per 1000 births)' on by one columns

In [41]:
# Build OLS model after multi-collinearity is addressed
inpc = sm.add_constant(inpv3)
ols = sm.OLS(endog=out, exog=inpc)
ols_mod = ols.fit()
ols_mod.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:     GDP ($ per capita)   R-squared:                       0.798
Model:                            OLS   Adj. R-squared:                  0.785
Method:                 Least Squares   F-statistic:                     59.78
Date:                Fri, 03 Sep 2021   Prob (F-statistic):           1.54e-65
Time:                        08:36:27   Log-Likelihood:                -2224.6
No. Observations:                 227   AIC:                             4479.
Df Residuals:                     212   BIC:                             4530.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const                         9620.3348    299.643     32.106      0.000    9029.673    1.02e+04
Population                    1.241e-06   3.87e-06      0.321      0.749   -6.38e-06    8.86e-06
Area_sqm                     -2.107e-05      0.000     -0.185      0.853      -0.000       0.000
Pop_Density_per sqm             -0.0026      0.186     -0.014      0.989      -0.370       0.365
Coastline (coast/area ratio)   -14.6493      3.764     -3.892      0.000     -22.068      -7.231
Net migration                  291.5991     45.495      6.409      0.000     201.918     381.280
Literacy (%)                    10.2348     24.183      0.423      0.673     -37.434      57.904
Phones (per 1000)               32.6685      2.389     13.673      0.000      27.959      37.378
Crops (%)                       12.0265     27.673      0.435      0.664     -42.524      66.577
Other (%)                       32.4206     25.903      1.252      0.212     -18.640      83.481
Climate                       -161.8270    556.875     -0.291      0.772   -1259.549     935.895
Birthrate                      -70.4072     56.913     -1.237      0.217    -182.595      41.781
Deathrate                       46.7923     61.463      0.761      0.447     -74.364     167.949
Agriculture                  -7980.2852   3201.322     -2.493      0.013   -1.43e+04   -1669.784
Service                      -2475.4906   2929.290     -0.845      0.399   -8249.756    3298.775
==============================================================================
Omnibus:                        9.575   Durbin-Watson:                   1.811
Prob(Omnibus):                  0.008   Jarque-Bera (JB):               15.191
Skew:                           0.227   Prob(JB):                     0.000503
Kurtosis:                       4.183   Cond. No.                     1.09e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.09e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

##### Interpretation
R**